### PR TITLE
Add unit test for Mixtral MoE layer

### PR DESCRIPTION
**Pull request by pcmoritz**
This is PR adds a unit test for the Mixtral MoE layer to vLLM.

It is based on @casper-hansen 's test in https://github.com/casper-hansen/AutoAWQ/blob/mixtral_fused/tests/test_fused_moe.py

### CODE REVIEW

File: vllm/model_executor/models/mixtral.py [hunk 1]
```diff
@@ -141,8 +142,9 @@ def forward(self, hidden_states: torch.Tensor) -> torch.Tensor:
                                         selected_experts,
                                         inplace=True)
 
-        final_hidden_states = tensor_model_parallel_all_reduce(
-            final_hidden_states)
+        if self.tp_size > 1:
```
Comment (pcmoritz) [inline vllm/model_executor/models/mixtral.py:145]: I'm curious what the thoughts on this are. On the one hand there is a ton of value in being able to run all the layers in a single process without having to stand up the distributed environment, on the other hand, this is redundant with the check in `tensor_model_parallel_all_reduce` that checks for `get_tensor_model_parallel_world_size() == 1` (but we can't use that here since `get_tensor_model_parallel_world_size` already needs the distributed environment. I considered monkey patching that function but that doesn't seem great either.

**Comment by simon-mo** [comment]
https://buildkite.com/vllm/ci/builds/744#018d5d59-27fa-4a6c-bc07-f15bd8d544fc/51-61
Looks like they failed. The docker image is public, you can reproduce them by running on L4 instances in GCP.

**Comment by pcmoritz** [comment]
The problem here is that pytorch and forking are fundamentally not compatible, and pytest-forked uses forking, so we can't run two pytorch tests in the same pytest-forked session. How do you feel about using pytest-xdist instead of pytest-forked? I think it uses "spawn" instead of "fork" under the hood so it works (I tried it). That's what pytorch seems to be doing for their own tests. The downside would be developers need to learn switching from `--forked` to `-n 8` or similar.

**Comment by pcmoritz** [comment]
I understand the problem, it comes from using pytorch with CUDA and pytest-forked. I think the best solution (short of using pytest-xdist instead of pytest-forked, which is a big change) is to isolate the layer wise tests from the model test and just use pytest without --forked for the layer wise tests (since these should be fast).

**Comment by simon-mo** [comment]
I’m fine completely not using forked here as well. 

On January 30, 2024, GitHub ***@***.***> wrote:
> I understand the problem, it comes from using pytorch with CUDA and
> pytest-forked. I think the best solution (short of using pytest-xdist
> instead of pytest-forked, which is a big change) is to isolate the
> layer wise tests from the model test and just use pytest without --
> forked for the layer wise tests (since these should be fast).
>

—
Reply to this email directly, view it on GitHub
<https://github.com/vllm-project/vllm/pull/2677#issuecomment-
1918344992>, or unsubscribe
<https://github.com/notifications/unsubscribe-
auth/AFBD7A6OES4M5G43M3XNEBTYRG7TPAVCNFSM6AAAAABCSISFS6VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTSMJYGM2DIOJZGI>.
You are receiving this because you commented.Message ID: <vllm-
***@***.***>

**Comment by pcmoritz** [comment]
Ok sounds good let's do that then. If it gets too slow going forward, we can switch to pytest-xdist :)

**Comment by pcmoritz** [comment]
I also added the end-to-end test for mixtral

**Comment by pcmoritz** [comment]
I had to shift it back to `layers`. The `models` tests don't like it if somebody else is using their precious GPU memory :D

**Comment by casper-hansen** [comment]
I’m just curious, is there any way this implementation can run FP16 and match the original implementation without a large difference in logits?

If not, do we understand why running both in FP16 gives such a large difference?

**Comment by pcmoritz** [comment]
Actually, in this test it is running in bfloat16 and the difference is not that large -- it is larger than without tensor cores though, but that's somewhat fundamental because tensor cores operate at less precision (see https://pytorch.org/docs/stable/notes/cuda.html#tf32-on-ampere). If you want the performance, you have to sacrifice the accuracy. In the future we will probably offer the option to do this arithmetic in fp8, that will be even less accurate and then we have to be super careful to do the right scaling.

I don't think this algorithm is fundamentally less numerically stable than what we had before, it is all just a big blocked matrix multiplication. The only difference is this one is more cache efficient (and does less work because we are only doing the work for the experts that are actually used). My guess is if you set the `allow_tf32=False`, the accuracy will be about the same.

**Comment by pcmoritz** [comment]
Ah correction, I had different versions of the test before. In this version it is float32, but here are the numerical differences for different dtypes (the maximum absolute difference of the states after the MOE layer):
```
tensor(0.0005, device='cuda:0', dtype=torch.float16)
tensor(0.0029, device='cuda:0', dtype=torch.bfloat16)
tensor(0.0008, device='cuda:0')
```

**Comment by pcmoritz** [comment]
Btw, one more point of comparison: Even if you trust the HuggingFace implementation a lot, look at the difference of evaluating the float32 version of it in on float16 vs. float32 inputs:

```python
import torch
from transformers import MixtralConfig
from transformers.models.mixtral.modeling_mixtral import MixtralSparseMoeBlock


config = MixtralConfig()
hf_moe = MixtralSparseMoeBlock(config).to("cuda")

inputs = torch.randn((1, 16, config.hidden_size)).to("cuda")

hf_states1, _ = hf_moe.forward(inputs)
hf_states2, _ = hf_moe.to(torch.float16).forward(inputs.to(torch.float16))

print("diff", torch.max(abs(hf_states1 - hf_states2.to(torch.float32))))
```

diff tensor(0.0003, device='cuda:0', grad_fn=<MaxBackward1>)

And this is almost the best possible case -- if you compare float16 weights on float16 inputs with bfloat16 weights on bfloat16 inputes, you get much worse (namely 0.0032 error). So we can actually be pretty happy about the accuracy we are getting.

**Comment by pcmoritz** [comment]
@casper-hansen I have also added tests for the other dtypes now :)

**Review by simon-mo — APPROVED** [review]
